wgengine/magicsock/magicsock_linux.go

```diff
@@ -0,0 +1,316 @@
+})()
+
+// These are our BPF filters that we use for testing packets.
+var (
+	// These were adapted from filters generated using tcpdump:
+	//    tcpdump -i lo -d 'udp[8:4]==0x5453f09f && udp[12:2]==0x92ac'
+	magicsockFilterV4 = []bpf.Instruction{
+		// Check Ethernet header for EtherType = 0x0800 (IPv4)
+		bpf.LoadAbsolute{Off: 12, Size: 2},
+		bpf.JumpIf{Cond: bpf.JumpEqual, Val: 0x800, SkipTrue: 0, SkipFalse: 10},
+
+		// Check protocol == UDP
+		bpf.LoadAbsolute{Off: ethHeaderSize + 9, Size: 1},
+		bpf.JumpIf{Cond: bpf.JumpEqual, Val: uint32(ipproto.UDP), SkipTrue: 0, SkipFalse: 8},
+
+		// Check for non-fragmented packets
+		bpf.LoadAbsolute{Off: ethHeaderSize + 6, Size: 2},
+		bpf.JumpIf{Cond: bpf.JumpBitsSet, Val: 0x1fff, SkipTrue: 6, SkipFalse: 0},
+
+		// Load IP header length into X
+		bpf.LoadMemShift{Off: ethHeaderSize + 0},
+
+		// Get the first 4 bytes of the UDP packet, compare with our magic number
+		bpf.LoadIndirect{Off: ethHeaderSize + udpHeaderSize, Size: 4},
+		bpf.JumpIf{Cond: bpf.JumpEqual, Val: magic1, SkipTrue: 0, SkipFalse: 3},
+
+		// Compare the next 2 bytes
+		bpf.LoadIndirect{Off: ethHeaderSize + udpHeaderSize + 4, Size: 2},
+		bpf.JumpIf{Cond: bpf.JumpEqual, Val: uint32(magic2), SkipTrue: 0, SkipFalse: 1},
+
+		// Accept the whole packet
+		bpf.RetConstant{Val: 0xFFFFFFFF},
+
+		// Skip the packet
+		bpf.RetConstant{Val: 0x0},
+	}
+
+	// IPv6 is more complicated to filter, since we can have 0-to-N
+	// extension headers following the IPv6 header. Since BPF can't loop,
+	// we can't really parse these in a general way; instead, we simply
+	// handle the case where we have either no extension headers, or a
+	// fragment header. Anything else will be skipped.
+	//
+	// The "proper" way to handle this would be to do minimal parsing in
+	// BPF and more in-depth parsing of all IPv6 packets in userspace, but
+	// on systems with a high volume of UDP that would be unacceptably slow
+	// and thus we'd rather be conservative here and possibly not receive
+	// disco packets rather than slow down the system.
+	magicsockFilterV6 = []bpf.Instruction{
+		// Check Ethernet header for EtherType = 0x86DD (IPv6)
+		bpf.LoadAbsolute{Off: 12, Size: 2},
+		bpf.JumpIf{Cond: bpf.JumpEqual, Val: 0x86DD, SkipTrue: 1, SkipFalse: 0},
+		bpf.RetConstant{Val: 0x0},
+
+		// Check "Next Header" field == UDP
+		bpf.LoadAbsolute{Off: ethHeaderSize + 6, Size: 1},
+		bpf.JumpIf{Cond: bpf.JumpEqual, Val: uint32(ipproto.UDP), SkipTrue: 0, SkipFalse: 1},
+
+		// Next header indicates we have a UDP packet, so the header
+		// offset is immediately following the IPv6 header.
+		bpf.LoadConstant{Dst: bpf.RegX, Val: ethHeaderSize + ipv6.HeaderLen},
+		bpf.Jump{Skip: 6}, // move past fragment checks
+
+		// Check if the "Next Header" field is the "Fragment" header
+		bpf.JumpIf{Cond: bpf.JumpEqual, Val: 0x2C, SkipTrue: 1, SkipFalse: 0},
+		bpf.RetConstant{Val: 0x0}, // skip packet
+
+		// If it's the fragment header, check whether the protocol is UDP
```

Comment:
You could almost certainly also decline to handle fragmented IPv6 packets. IPv6 strongly states that fragmentation is understanding-optional in IPv6, meaning that most of the internet doesn't generate any.

But if you do handle it, you need to add an extra check to only process packets with fragment offset == 0, otherwise you'll have false positives on later fragments.